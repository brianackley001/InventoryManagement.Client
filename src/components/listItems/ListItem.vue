<template>
  <div class="list-item-container">
    <v-list-item>
      <template>
        <v-list-item-action>
          <v-checkbox
            :input-value="displaySelectionCheckBox"
            v-show="itemIsSelectionEligible"
            dense
            @click="
              if (itemIsSelectionEligible) {
                trackSelectionMode({
                  primaryKeyId: $route.params.id || 0,
                  valueId: itemId,
                });
              }
            "
          ></v-checkbox>
          <v-icon
            left
            small
            v-text="searchResultNonItemClass"
            v-show="!itemIsSelectionEligible"
          ></v-icon>
        </v-list-item-action>
        <v-list-item-content>
          <v-list-item-title
            v-text="item.name"
            @click="launchModal"
          ></v-list-item-title>
          <v-list-item-subtitle v-if="item.groups && item.groups.length > 0">
            <!-- <v-icon left small>mdi-group</v-icon> -->
            <AttributeChip
              v-for="(g, index) in item.groups"
              :key="index"
              :attributeType="enums.ATTRIBUTE_TYPE.GROUP"
              :item="g"
              isListItem
              readOnly
            />
          </v-list-item-subtitle>
          <v-list-item-subtitle v-if="item.tags && item.tags.length > 0">
            <!-- <v-icon left small>mdi-tag</v-icon> -->
            <AttributeChip
              v-for="(t, index) in item.tags"
              :key="index"
              :attributeType="enums.ATTRIBUTE_TYPE.TAG"
              :item="t"
              isListItem
              readOnly
            />
          </v-list-item-subtitle>
        </v-list-item-content>

        <v-list-item-avatar @click="launchModal">
          <v-icon
            v-show="itemIsSelectionEligible"
            :color="iconColor"
            :class="thumbColor"
            v-text="iconClass"
          ></v-icon>
        </v-list-item-avatar>
      </template>
    </v-list-item>
    <v-divider></v-divider>
  </div>
</template>

<script>
import { mapActions, mapGetters } from "vuex";
import enumValues from "@/assets/configuration/enums";
import AttributeChip from "@/components/AttributeChip";
import utils from "@/utils/attributeListItem";
export default {
  name: "listItem",
  props: ["attributeType", "item"],
  components: { AttributeChip },
  computed: {
    ...mapGetters("selections", ["selectionMode", "selectionModeItems"]),
    displaySelectionCheckBox: function () {
      return this.selectionModeItems.indexOf(this.itemId) > -1;
    },
    enums: function () {
      return enumValues;
    },
    iconColor: function () {
      return this.item.amountValue < 1
        ? this.enums.ITEM_QUANTITY.ICON_COLOR.EMPTY
        : this.item.amountValue == 1
        ? this.enums.ITEM_QUANTITY.ICON_COLOR.LOW
        : this.enums.ITEM_QUANTITY.ICON_COLOR.FULL;
    },
    iconClass: function () {
      return this.item.amountValue < 1
        ? this.enums.ITEM_QUANTITY.ICON_CLASS.EMPTY
        : this.item.amountValue == 1
        ? this.enums.ITEM_QUANTITY.ICON_CLASS.LOW
        : this.enums.ITEM_QUANTITY.ICON_CLASS.FULL;
    },
    isCurrentRoute: function () {
      return (
        this.$route.name ===
          utils.getAttributeTypeRouteName(this.attributeType) &&
        this.$route.params.id === this.item.id
      );
    },
    itemId: function () {
      return this.attributeType == this.enums.ATTRIBUTE_TYPE.SHOPPING_LIST
        ? this.item.itemId
        : this.item.id;
    },
    itemIsSelectionEligible: function () {
      return (
        (this.attributeType == this.enums.ATTRIBUTE_TYPE.SEARCH &&
          this.item.resultType == this.enums.ATTRIBUTE_TYPE.ITEM) ||
        this.attributeType !== this.enums.ATTRIBUTE_TYPE.SEARCH
      );
    },
    searchResultNonItemClass: function () {
      switch (this.item.resultType) {
        case this.enums.ATTRIBUTE_TYPE.TAG:
          return "mdi-tag";
        case this.enums.ATTRIBUTE_TYPE.SHOPPING_LIST:
          return "mdi-cart";
        default:
          return "mdi-group";
      }
    },
    thumbColor: function () {
      return this.$vuetify.theme.dark
        ? this.enums.ITEM_QUANTITY.THUMB_COLOR.DARK
        : this.enums.ITEM_QUANTITY.THUMB_COLOR.LIGHT;
    },
  },
  data: function () {
    return {};
  },
  methods: {
    ...mapActions("items", ["fetchItemDetail", "setDialogVisible"]),
    ...mapActions("selections", ["resetSelectionMode", "trackSelectionMode"]),
    isSearchByAttributeId(attributeItem, attributeType) {
      return (
        this.$route.name === attributeType &&
        attributeItem.id === this.$route.params.id
      );
    },
    launchModal() {
      // modal for ITEM type:
      if (this.itemIsSelectionEligible) {
        const fetchItemId =
          this.attributeType &&
          this.attributeType == this.enums.ATTRIBUTE_TYPE.SHOPPING_LIST
            ? this.item.itemId
            : this.item.id;
        this.fetchItemDetail(fetchItemId);
        this.setDialogVisible(true);
        return;
      }

      // for non ITEM types in search results, direct to Attribute Type page (e.g. Group, Tag, Shopping List...)
      if (!this.isCurrentRoute) {
        utils.navigateToAttributeCollection(
          this.item.resultType,
          this.item.id,
          this.item.name
        );
      }
    },
  },
};
</script>

<style lang="scss"></style>
