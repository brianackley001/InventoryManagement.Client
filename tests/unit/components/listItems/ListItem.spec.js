// Libraries
//import Vue from "vue";
import Vuex from "vuex";
import Vuetify from "vuetify";
// Components
import ListItem from "@/components/listItems/ListItem.vue";
// Utilities
import { mount, createLocalVue } from "@vue/test-utils";

// NOTE - console errors until vuetify 3.x: https://stackoverflow.com/questions/51990753/vue-js-vuetify-issue-running-my-first-unit-test-with-jest
// https://github.com/vuetifyjs/vuetify/issues/4964
describe("ListItem - Read Mode", () => {
  const localVue = createLocalVue();
  let wrapper;
  let objListItem;

  beforeEach(() => {
    localVue.use(Vuex);
    localVue.use(Vuetify);
    objListItem = {
      id: 10101,
      name: "Test Name",
      description: "Test Description summary text",
      sku: "1234567",
      price: 9.99,
      salePrice: 4.44,
      pricePerUnit: 1.18,
      url: null,
      imageUrl: null,
      isActive: true,
      createDate: null,
      updateDate: null,
      vendor: null,
      amountRemaining: 0,
      tags: ["Food"],
      groups: ["Costco"],
    };
    wrapper = mount(ListItem, {
      localVue,
      propsData: {
        item: objListItem,
      },
      mocks: {
        $store: {
          state: { selectionModeItems: [999, 16098] },
          dispatch: jest.fn(),
        },
      },
    });
  });

  it("should match snapshot", () => {
    // With jest we can create snapshot files of the HTML output
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("should render the item name as title", () => {
    const selector = wrapper.find(".v-list-item__title");
    expect(selector.text()).toBe(objListItem.name);
  });
  it("should render the description as subtitle", () => {
    const selector = wrapper.find(".v-list-item__subtitle");
    expect(selector.text().trim()).toMatch(objListItem.description);
  });
  it("should render amount icon in default view", () => {
    //  With beforeEach assignment ( amountRemaining: 0 ) - the "empty" icon will be present. Need global CSS variables to test against
    const selector = wrapper.find(
      ".v-avatar.v-list-item__avatar > i.v-icon.notranslate.white.mdi.mdi-gauge-empty"
    );
    expect(selector.exists()).toBe(true);
  });
  //   it("should toggle amount icon to checkbox icon on click", () => {
  //     //const event = jest.fn();
  //     const selector = wrapper.find(".v-avatar.v-list-item__avatar ");
  //     // Simulate a click on the button
  //     selector.trigger("click");
  //     expect(selector.text().trim()).toMatch(objListItem.description);
  //   });
});
// div.v-avatar.v-list-item__avatar > i.v-icon.notranslate.white.mdi.mdi-gauge-low.theme--light.amber--text.text--darken-1
// div.v-avatar.v-list-item__avatar > i.v-icon.notranslate.white.mdi.mdi-gauge-empty.theme--light.red--text.text--darken-1
// div.v-avatar.v-list-item__avatar > i.v-icon.notranslate.white.mdi.mdi-gauge-full.theme--light.green--text

/*
TO-DO:
    Global CSS Values or app settings for testing CSS Variables/colors/icons
    Click Events will need a real store
    Shopping List Item view/behavior
    Test coverage results will most likely tank without supressing the "[Vuetify] Multiple instances of Vue detected" error 
*/
