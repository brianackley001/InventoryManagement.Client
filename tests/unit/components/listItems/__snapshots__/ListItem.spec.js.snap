// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem - Read Mode should match snapshot 1`] = `
<div class="list-view-item-container">
  <div tabindex="-1" class="v-list-item theme--light">
    <div class="v-avatar v-list-item__avatar" style="height: 40px; min-width: 40px; width: 40px;"><i aria-hidden="true" class="v-icon notranslate white mdi mdi-gauge-empty theme--light red--text text--darken-1"></i> <i aria-hidden="true" class="v-icon notranslate light-blue lighten-5 mdi mdi-check-circle-outline theme--light" style="display: none;"></i></div>
    <div class="v-list-item__content">
      <div class="v-list-item__title">Test Name</div>
      <div class="v-list-item__subtitle"><span><span item="Costco" class="group">Costco</span> <span item="Food" class="tag">Food</span></span>
        Test Description summary text
      </div>
    </div>
  </div>
  <hr role="separator" aria-orientation="horizontal" class="v-divider v-divider--inset theme--light">
</div>
`;

exports[`ListItem.vue should match snapshot 1`] = `
<div class="list-view-item-container">
  <div tabindex="-1" class="v-list-item theme--light">
    <div class="v-avatar v-list-item__avatar" style="height: 40px; min-width: 40px; width: 40px;"><i aria-hidden="true" class="v-icon notranslate white mdi mdi-gauge-empty theme--light red--text text--darken-1"></i> <i aria-hidden="true" class="v-icon notranslate light-blue lighten-5 mdi mdi-check-circle-outline theme--light" style="display: none;"></i></div>
    <div class="v-list-item__content">
      <div class="v-list-item__title">Test Name</div>
      <div class="v-list-item__subtitle"><span><span item="Costco" class="group">Costco</span> <span item="Food" class="tag">Food</span></span>
        Test Description summary text
      </div>
    </div>
  </div>
  <hr role="separator" aria-orientation="horizontal" class="v-divider v-divider--inset theme--light">
</div>
`;
